src/group_theory/is_free_group.lean

```diff
@@ -18,47 +34,74 @@ universes u w
 class is_free_group (G : Type u) [group G] : Type (u+1) :=
 (generators : Type u)
 (of : generators → G)
-(unique_lift : ∀ {X : Type u} [group X] (f : generators → X),
+(unique_lift' : ∀ {X : Type u} [group X] (f : generators → X),
                 ∃! F : G →* X, ∀ a, F (of a) = f a)
 
 instance free_group_is_free_group {A} : is_free_group (free_group A) :=
 { generators := A,
   of := free_group.of,
-  unique_lift := begin
+  unique_lift' := begin
     introsI X _ f,
-    exact ⟨free_group.lift f,
-      λ _, free_group.lift.of,
-      λ g hg, monoid_hom.ext (λ _, free_group.lift.unique g hg)⟩
+    have := free_group.lift.symm.bijective.exists_unique f,
+    simp_rw function.funext_iff at this,
+    exact this,
   end }
 
 namespace is_free_group
 
 variables {G H : Type u} [group G] [group H] [is_free_group G]
 
-/-- The group homomorphism from a free group given by some function on the generators. -/
-def lift' (f : generators G → H) : G →* H := classical.some (unique_lift f)
+/-- The equivalence between functions on the generators and group homomorphisms from a free group
+given by those generators. -/
+@[simps symm_apply]
+def lift' : (generators G → H) ≃ (G →* H) :=
+{ to_fun := λ f, classical.some (unique_lift' f),
+  inv_fun := λ F, F ∘ of,
+  left_inv := λ f, funext (classical.some_spec (unique_lift' f)).left,
+  right_inv := λ F, ((classical.some_spec (unique_lift' (F ∘ of))).right F (λ _, rfl)).symm }
+
 @[simp] lemma lift'_of (f : generators G → H) (a : generators G) : (lift' f) (of a) = f a :=
-(classical.some_spec (unique_lift f)).left a
+congr_fun (lift'.symm_apply_apply f) a
+
+@[simp] lemma lift'_eq_free_group_lift {A : Type u} :
+  (@lift' (free_group A) G _ _ _) = free_group.lift :=
+begin
+  -- TODO: `apply equiv.symm_bijective.injective`,
+  rw [←free_group.lift.symm_symm, ←(@lift' (free_group A) G _ _ _).symm_symm],
+  congr' 1,
+  ext,
+  refl,
+end
+
+@[simp] lemma of_eq_free_group_of {A : Type u} : (@of (free_group A) _ _) = free_group.of :=
+rfl
+
 lemma eq_lift' (f : generators G → H) (F : G →* H) (h : ∀ a, F (of a) = f a) : F = lift' f :=
-(classical.some_spec (unique_lift f)).right F h
+lift'.symm_apply_eq.mp (funext h)
+
+@[ext]
+lemma ext_hom ⦃f g : G →* G⦄ (h : ∀ a, f (of a) = g (of a)) :
```

Comment:
Can we generalise so `f` and `g` go into any group, possibly in another universe?

Reply:
Not here, because this lemma is used to prove `lift`. I'll rename this to `ext_hom'`, and then create an unprimed version

Reply:
Ah, I missed this was two requests. I've generalized this for `G →* H`, and added a new lemma for `G →* X`